finality-aleph/src/metrics/chain_state.rs

```diff
@@ -91,47 +133,105 @@ pub async fn run_chain_state_metrics<
         }
     };
 
+    let mut cache: LruCache<TP::TxHash, Instant> = LruCache::new(
+        NonZeroUsize::new(TRANSACTION_CACHE_SIZE).expect("the cache size is a non-zero constant"),
+    );
+
     let mut best_block_notifications = import_notifications
         .fuse()
         .filter(|notification| future::ready(notification.is_new_best));
     let mut finality_notifications = finality_notifications.fuse();
-
     let mut previous_best: Option<HE> = None;
+
     loop {
-        select! {
-            maybe_block = best_block_notifications.next() => {
-                match maybe_block {
-                    Some(block) => {
-                        let number = (*block.header.number()).saturated_into::<BlockNumber>();
-                        metrics.update_best_block(number);
-                        if let Some(reorg_len) = detect_reorgs(backend, previous_best, block.header.clone()) {
-                            metrics.report_reorg(reorg_len);
-                        }
-                        previous_best = Some(block.header);
+        iteration(
+            backend,
+            &mut best_block_notifications,
+            &mut finality_notifications,
+            &metrics,
+            &mut transaction_pool_info_provider,
+            &mut cache,
+            &mut previous_best,
+        )
+        .await;
+    }
+}
+
+async fn iteration<
+    X,
+    HE: HeaderT<Number = BlockNumber, Hash = B::Hash>,
+    B: BlockT<Header = HE, Extrinsic = X>,
+    BE: HeaderMetadata<B> + BlockBackend<B>,
+    TP: TransactionPoolInfoProvider<Extrinsic = X>,
+    BBN: Stream<Item = BlockImportNotification<B>> + Unpin,
+    FN: Stream<Item = FinalityNotification<B>> + Unpin,
+>(
+    backend: &BE,
+    best_block_notifications: &mut BBN,
+    finality_notifications: &mut FN,
+    metrics: &ChainStateMetrics,
+    transaction_pool_info_provider: &mut TP,
+    cache: &mut LruCache<TP::TxHash, Instant>,
+    previous_best: &mut Option<HE>,
+) {
+    select! {
+        maybe_block = best_block_notifications.next() => {
+            match maybe_block {
+                Some(block) => {
+                    let number = (*block.header.number()).saturated_into::<BlockNumber>();
```

Comment:
Good catch, it was needed previously because I was too lazy to write the correct equations on traits, but now it can be removed (as well as the other saturated_into in that file for reorg detection)